agent/app/agent.go

```diff
@@ -377,6 +396,44 @@ func (agent *ecsAgent) doStart(containerChangeEventStream *eventstream.EventStre
 		deregisterInstanceEventStream, client, state, taskHandler)
 }
 
+// waitUntilInstanceInService Polls IMDS until the target lifecycle state indicates that the instance is going in
+// service. This is to avoid instances going to a warm pool being registered as container instances with the cluster
+func (agent *ecsAgent) waitUntilInstanceInService() error {
+	var err error
+	var targetState string
+	// Poll while the instance is in a warmed state or while waiting for the data to be populated.
+	// If the data is not populated after a certain number of polls, then stop polling and return the not found error.
+	// The polling maximum does not apply to instances in the warmed states
+	for i := 0; i < asgLifeCyclePollMax || targetState != ""; i++ {
+		targetState, err = agent.getTargetLifeCycle()
+		// stop polling if the retrieved state is in service or we get an unexpected error
+		if (targetState == inServiceState) || (err != nil && !strings.Contains(err.Error(), "status code: 404")) {
```

Comment:
Could we check if err.Code() returns the error code?

Reply:
I don't think Code() is part of the error interface, just the error strings. And IMDS returns the HTML, which isn't ideal for checking details like this. I hate to rely on string contains for this, but I think trying to parse the error would involve more hardcoding and unnecessary extra work